### DISCUSSION TITLE
What is the fastest way to convert `image-rs::Image` into `slint::Image`?

### DISCUSSION THREAD

**Discussion by qarmin** [slint-ui/slint#3912] [Ideas]
Currently I'm using code similar to

```
pub type ImageBufferRgba = image::ImageBuffer<image::Rgba<u8>, Vec<u8>>;

fn convert_into_slint_image(img: DynamicImage) -> slint::Image {
    let image_buffer: ImageBufferRgba = img.to_rgba8();
    let buffer = slint::SharedPixelBuffer::<slint::Rgba8Pixel>::clone_from_slice(image_buffer.as_raw(), image_buffer.width(), image_buffer.height());
    slint::Image::from_rgba8(buffer)
}

let img = image::open(image_name).unwrap();
let img2 = convert_into_slint_image(img);
app.global::<GuiState>().set_preview_image(slint_image);
```

In debug build, jpg image 5184x3495(~18 million pixels) which weight is 19.1 MB loads in ~1.1 second
```
Loading image took: 374.248528ms, converting image took: 782.257233ms, setting image took: 4.586µs
```
In release build it takes lower amount of time
```
Loading image took: 367.533911ms, converting image took: 85.82906ms, setting image took: 1.47µs
```

Is there any way to decrease convert time between formats especially in debug builds?

**Comment by ogoffart** [comment]
The best would be to avoid conversion altogether.  (ie: trying to always operate with a SharedPixelBuffer.

But there might be optimizations to do in `clone_from_slice()`  The current implementation:

https://github.com/slint-ui/slint/blob/df7657dc2d1fb37f17ff0f7285af68bf6704fd22/internal/core/graphics/image.rs#L129-L139

Maybe `.iter().cloned().collect()` is not ideal especially debug mode.  Maybe something like this would work

```rust
Vec::from(pixel_slice.as_pixels())
```

and would be more optimized in debug mode.

**Reply by ogoffart** [reply]
I pushed that change in 1a0e065a5441d3c71ee8e8b21467014438cb95b1 which should in principle improve the debug build

**Reply by qarmin** [reply]
Before
```
Loading image took: 374.248528ms, converting image took: 782.257233ms, setting image took: 4.586µs
```
Now
```
Loading image took: 380.358729ms, converting image took: 488.832646ms, setting image took: 7.982µs
```

**Comment by tronical** [comment]
convert_into_slint_image() does two things:

(1) convert into a rgba8 image
(2) creating a copy via clone_from_slice

It would be interesting to know how long step (1) takes. Perhaps there's a way to make it so that the decoder creates rgba8 right away and so that this step vanishes.

For (2) - if it turns out to be costly - we would need a way adopt a Vec's data. That would require changing the SharedVector internals to allow for an indirection that we typically don't need.

Perhaps another option that might work with today's APIs would be to use image-rs lower level APIs to create a Reader, determine dimensions, allocate a SharedPixelBuffer and decode directly into it. Not sure if that's always possible or at what other cost this comes.

**Reply by qarmin** [reply]
DEBUG:
Converting image to rgba8 took: 43.705888ms
Converting slice into shared pixelbuffer took: 748.677265ms
Converting sharedpixelbuffer to slint image took: 416ns
Loading image took: 390.485237ms, converting image took: 800.394429ms, setting image took: 5.231µs

RELEASE:
Converting image to rgba8 took: 44.366056ms
Converting slice into shared pixelbuffer took: 35.693432ms
Converting sharedpixelbuffer to slint image took: 202ns
Loading image took: 352.566009ms, converting image took: 88.694365ms, setting image took: 1.305µs
